src/sage/schemes/elliptic_curves/homset.py

```diff
@@ -0,0 +1,234 @@
+            True
+            sage: Hom(EE0, EE1) in CommutativeAdditiveGroups()
+            True
+            sage: Hom(E0, E1) in Rings()
+            False
+            sage: Hom(EE0, EE1) in Rings()
+            False
+        """
+        if not isinstance(E1, EllipticCurve_generic):
+            raise ValueError('domain must be an elliptic curve')
+        if not isinstance(E2, EllipticCurve_generic):
+            raise ValueError('codomain must be an elliptic curve')
+        base = E1.base_ring()
+        if base != E2.base_ring():
+            raise ValueError('domain and codomain must have the same base ring')
+
+        super().__init__(E1, E2, category=category, base=base)
+
+        from sage.categories.commutative_additive_groups import CommutativeAdditiveGroups
+        self._refine_category_(CommutativeAdditiveGroups())
+
+        if self.is_endomorphism_set():
+            from sage.categories.rings import Rings
+            self._refine_category_(Rings())
+
+            assert E1 == E2
+
+            # The endomorphism ring of an elliptic curve over a finite field
+            # is non-commutative if and only if the curve is supersingular and
+            # the Frobenius endomorphism equals a scalar.
+            if (base.characteristic() == 0 or E1.is_ordinary() or
+                    (base.is_finite() and E1.frobenius() not in ZZ)):
+                from sage.categories.commutative_rings import CommutativeRings
+                self._refine_category_(CommutativeRings())
+
+    def _coerce_map_from_(self, other):
+        r"""
+        Check if this homset has a coercion map from another
+        parent ``other``.
+
+        The only currently supported case is when this homset has
+        equal domain and codomain: In this case elements from `\ZZ`
+        are embedded as scalar multiplications.
+
+        EXAMPLES::
+
+            sage: E1 = EllipticCurve(j=42)
+            sage: E2 = EllipticCurve(j=43)
+            sage: Hom(E1, E1)._coerce_map_from_(ZZ)
+            True
+            sage: Hom(E1, E2)._coerce_map_from_(ZZ)
+            False
+        """
+        return self.is_endomorphism_set() and other is ZZ
+
+    def _element_constructor_(self, data):
+        r"""
+        Construct an element of this homset from the given ``data``.
+
+        The only currently supported case is when this homset has
+        equal domain and codomain: In this case elements from `\ZZ`
+        are embedded as scalar multiplications.
+
+        EXAMPLES::
+
+            sage: E1 = EllipticCurve(j=42)
+            sage: E2 = EllipticCurve(j=43)
+            sage: Hom(E1, E1)(5)
+            Scalar-multiplication endomorphism [5] of Elliptic Curve defined by y^2 = x^3 + 5901*x + 1105454 over Rational Field
+            sage: Hom(E1, E2)(5)
+            Traceback (most recent call last):
+            ...
+            ValueError: domain and codomain must be equal
+        """
+        if self.codomain() != self.domain():
```

Comment:
```suggestion
        if not self.is_endomorphism_set():
```
For consistency, plus some protections for accidental deallocations.